code/game/objects/structures/props/beam_prism.dm

```diff
@@ -0,0 +1,218 @@
+
+	var/rotate_degrees = new_bearing - degrees_from_north
+
+	if(new_bearing == 360)
+		new_bearing = 0
+	degrees_from_north = new_bearing
+
+	var/two_stage = 0
+	if(rotate_degrees == 180 || rotate_degrees == -180)
+		two_stage = 1
+		var/multiplier = pick(-1, 1)
+		rotate_degrees = multiplier * (rotate_degrees / 2)
+
+	playsound(src, interaction_sound, 50, 1)
+	if(two_stage)
+		animate(src, transform = turn(src.transform, rotate_degrees), time = 3)
+		sleep(3)
+		animate(src, transform = turn(src.transform, rotate_degrees), time = 3)
+	else
+		animate(src, transform = turn(src.transform, rotate_degrees), time = 6)
+
+/obj/structure/prop/prism/bullet_act(var/obj/item/projectile/Proj)
+	if(istype(Proj, redirect_type))
+		visible_message("<span class='danger'>\The [src] redirects \the [Proj]!</span>")
+		flick("[initial(icon_state)]+glow", src)
+
+		var/new_x = (1 * round(10 * cos(degrees_from_north - 90))) + x //Vectors vectors vectors.
+		var/new_y = (-1 * round(10 * sin(degrees_from_north - 90))) + y
+		var/turf/curloc = get_turf(src)
+
+		Proj.penetrating += 1 // Needed for the beam to get out of the turret.
+
+		Proj.redirect(new_x, new_y, curloc, null)
+
+/obj/structure/prop/prism/incremental
+	free_rotate = 0
+	description_info = "This device is capable of redirecting any beam projectile, but only locks to specific positions in rotation."
+
+/obj/structure/prop/prism/incremental/externalcont
+	external_control_lock = 1
+	description_info = "This device is capable of redirecting any beam projectile, but only locks to specific positions in rotation when rotated by a control dial."
+
+/obj/structure/prop/prism/externalcont
+	external_control_lock = 1
+	description_info = "This device is capable of redirecting any beam projectile, but can only be rotated by an external control dial."
+
+/obj/structure/prop/prismcontrol
+	name = "prismatic dial"
+	desc = "A large dial with a crystalline ring."
+	icon = 'icons/obj/props/prism.dmi'
+	icon_state = "dial"
+	density = FALSE
+	anchored = TRUE
+
+	interaction_message = "<span class='notice'>The dial pulses as your hand nears it.</span>"
+	var/list/my_turrets = list()
+	var/dialID = null
+
+/obj/structure/prop/prismcontrol/attack_hand(mob/living/user)
+	..()
+
+	var/confirm = input("Do you want to try to rotate \the [src]?", "[name]") in list("Yes", "No")
+	if(confirm == "No")
+		visible_message(\
+			"<span class='notice'>[user.name] decides not to try turning \the [src].</span>",\
+			"<span class='notice'>You decide not to try turning \the [src].</span>")
+		return
+
+	if(!my_turrets || !my_turrets.len)
+		to_chat(user, "<span class='notice'>\The [src] doesn't seem to do anything.</span>")
+		return
+
+	var/free_rotate = 1
+	var/list/compass_directions = list()
+	for(var/obj/structure/prop/prism/P in my_turrets)
+		if(!P.free_rotate)
+			free_rotate = 0
+		else
+			free_rotate = 0
+			compass_directions = P.compass_directions
+			break
```

Comment:
For each prism I control, if it can only rotate to cardinals, then I can only rotate to cardinals. 
If it can be rotated to non-cardinals, then I can only be rotated to cardinals, and the directions I can rotate to are the cardinals
(Such that cardinals includes composites, like NW)

That's the description of this for loop. I don't think it's what you're intending, but I could be wrong